### PR TITLE
Fragmenting cylinder example

### PR DESCRIPTION
**Pull request by pabloseleson**
Adds new fragmentation example

### CODE REVIEW

**Comment by streeve** [comment]
@pabloseleson I got this partially working: I applied the force BC only on the inner layer of particles and get fragmentation as we expect. There is a mistake somewhere leading to a very obvious lack of symmetry

Other things before getting that to work:
- This should likely be a velocity BC instead 
- I added the dynamic local critical stretch and did not see a noticeable difference
- We can add contact later for a more physical problem

File: examples/mechanics/crack_branching.cpp [hunk 1]
```diff
@@ -90,13 +90,14 @@ void crackBranchingExample( const std::string filename )
...
     double b0 = sigma0 / dy;
 
-    CabanaPD::RegionBoundary plane1( low_corner[0], high_corner[0],
-                                     low_corner[1] - dy, low_corner[1] + dy,
-                                     low_corner[2], high_corner[2] );
-    CabanaPD::RegionBoundary plane2( low_corner[0], high_corner[0],
-                                     high_corner[1] - dy, high_corner[1] + dy,
-                                     low_corner[2], high_corner[2] );
-    std::vector<CabanaPD::RegionBoundary> planes = { plane1, plane2 };
+    CabanaPD::RegionBoundary<CabanaPD::Rectilinear> plane1(
```
Comment (pabloseleson) [inline examples/mechanics/crack_branching.cpp:93]: Why was crack branching changed for the fragmenting cylinder example?

File: examples/mechanics/fragmenting_cylinder.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    std::array<double, 3> high_corner = inputs["high_corner"];
+    std::array<int, 3> num_cells = inputs["num_cells"];
+    int m = std::floor( delta /
+                        ( ( high_corner[0] - low_corner[0] ) / num_cells[0] ) );
+    int halo_width = m + 1; // Just to be safe.
+
+    // ====================================================
+    //                    Force model
+    // ====================================================
+    using model_type = CabanaPD::ForceModel<CabanaPD::PMB, CabanaPD::Fracture,
```
Comment (pabloseleson) [inline examples/mechanics/fragmenting_cylinder.cpp:59]: We need to ensure the other examples are updated accordingly.

File: examples/mechanics/fragmenting_cylinder.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    //     CabanaPD::ForceModel<CabanaPD::LPS, CabanaPD::Fracture>;
+    // model_type force_model( delta, K, G, G0 );
+
+    // ====================================================
+    //                 Particle generation
+    // ====================================================
+    // Does not set displacements, velocities, etc.
+    auto particles = std::make_shared<
+        CabanaPD::Particles<memory_space, typename model_type::base_model,
+                            typename model_type::thermal_type>>(
```
Comment (pabloseleson) [inline examples/mechanics/fragmenting_cylinder.cpp:73]: Where is "thermal_type" defined?

File: examples/mechanics/crack_branching.cpp (hunk 1 above)
Comment (streeve) [reply examples/mechanics/crack_branching.cpp:93]: See changes in the BC file - there was previous no cylindrical option, so the naming/setup needed to be changed. That can be moved to a new PR, but only if necessary

File: examples/mechanics/fragmenting_cylinder.cpp (hunk 2 above)
Comment (streeve) [reply examples/mechanics/fragmenting_cylinder.cpp:59]: I don't think so - there are defaults to avoid that

File: examples/mechanics/fragmenting_cylinder.cpp (hunk 3 above)
Comment (streeve) [reply examples/mechanics/fragmenting_cylinder.cpp:73]: Inside each model

File: src/CabanaPD_Boundary.hpp [hunk 4]
```diff
@@ -21,8 +21,20 @@
...
+};
+
+// Forward declaration.
+template <class GeometryType>
+struct RegionBoundary;
+
 // Define a plane or other rectilinear subset of the system as the boundary.
-struct RegionBoundary
+template <>
+struct RegionBoundary<Rectilinear>
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:37]: Should "rectilinear" be instead "Rectangular Prism" or so?

File: src/CabanaPD_Boundary.hpp [hunk 5]
```diff
@@ -21,8 +21,20 @@
...
+};
+struct Cylinder
+{
+};
+
+// Forward declaration.
+template <class GeometryType>
+struct RegionBoundary;
+
 // Define a plane or other rectilinear subset of the system as the boundary.
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:35]: This is not strictly speaking a plane. Also, what does it mean "or other rectilinear subset"? It seems this is specific to a rectangular prism.

File: src/CabanaPD_Boundary.hpp [hunk 6]
```diff
@@ -97,9 +149,7 @@ struct BoundaryIndexSpace<MemorySpace, RegionBoundary>
         Kokkos::RangePolicy<ExecSpace> policy( 0, particles.n_local );
         auto index_functor = KOKKOS_LAMBDA( const std::size_t pid )
         {
-            if ( x( pid, 0 ) >= plane.low_x && x( pid, 0 ) <= plane.high_x &&
-                 x( pid, 1 ) >= plane.low_y && x( pid, 1 ) <= plane.high_y &&
-                 x( pid, 2 ) >= plane.low_z && x( pid, 2 ) <= plane.high_z )
+            if ( region.compare( x, pid ) )
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:152]: What does this line do? Compare with what?

File: src/CabanaPD_Boundary.hpp [hunk 7]
```diff
@@ -288,7 +335,7 @@ auto createBoundaryCondition( BCTag, const double value, ExecSpace exec_space,
 {
     using memory_space = typename Particles::memory_space;
     using bc_index_type = BoundaryIndexSpace<memory_space, BoundaryType>;
-    bc_index_type bc_indices = createBoundaryIndexSpace<BoundaryType>(
+    bc_index_type bc_indices = createBoundaryIndexSpace(
         exec_space, particles, planes, initial_guess );
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:339]: What is "initial_guess" for?

File: src/force/CabanaPD_ForceModels_PMB.hpp [hunk 8]
```diff
@@ -91,10 +91,75 @@ struct ForceModel<PMB, Fracture, TemperatureIndependent>
...
+
+    // Use the dynamic s0 from the minLocalStretch function, reused through the
+    // neighbor loop.
     KOKKOS_INLINE_FUNCTION
-    bool criticalStretch( const int, const int, const double r,
-                          const double xi ) const
+    bool criticalStretch( const int, const int, const double r, const double xi,
+                          const double s0_i ) const
     {
+        double bond_break_coeff = ( 1.0 + s0_i ) * ( 1.0 + s0_i );
```
Comment (pabloseleson) [inline src/force/CabanaPD_ForceModels_PMB.hpp:162]: In the dynamic stretch, to preserve symmetry, the condition for bond breaking should depend on both i and j. 

In PDLAMMPS, they have:

if (stretch > MIN(s0[i],s0[j])) partner[i][jj] = 0;

So, the minimum of the two critical stretched is used.

Also, we need to ensure the critical stretch for each particle is computed before breaking any bonds. Otherwise, some will account more bonds than other.

File: src/force/CabanaPD_ForceModels_PMB.hpp [hunk 9]
```diff
@@ -231,7 +296,7 @@ struct ForceModel<PMB, Fracture, TemperatureDependent, TemperatureType>
 
     KOKKOS_INLINE_FUNCTION
     bool criticalStretch( const int i, const int j, const double r,
-                          const double xi ) const
+                          const double xi, const double ) const
```
Comment (pabloseleson) [inline src/force/CabanaPD_ForceModels_PMB.hpp:299]: What is this change for?

File: src/CabanaPD_Boundary.hpp (hunk 6 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:152]: Will change to `inside()`

**Comment by pabloseleson** [comment]
Current simulation:

https://github.com/user-attachments/assets/1b341455-7cd0-4935-8134-5262b354a7c8

File: examples/mechanics/fragmenting_cylinder.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    std::array<double, 3> system_size = inputs["system_size"];
+
+    double Rin = inputs["cylinder_inner_radius"];
+    double Rout = inputs["cylinder_outer_radius"];
+
+    double length = system_size[0];
+    double width = system_size[1];
+    double height = system_size[2];
+
+    // Define empty boundary region
```
Comment (pabloseleson) [inline examples/mechanics/fragmenting_cylinder.cpp:86]: We need to remove the boundary condition here.

File: examples/mechanics/fragmenting_cylinder.cpp [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    int m = std::floor( delta /
+                        ( ( high_corner[0] - low_corner[0] ) / num_cells[0] ) );
+    int halo_width = m + 1; // Just to be safe.
+
+    // ====================================================
+    //                    Force model
+    // ====================================================
+    using model_type = CabanaPD::ForceModel<CabanaPD::PMB, CabanaPD::Fracture>;
+    model_type force_model( delta, K, G0 );
+    // using model_type =
```
Comment (pabloseleson) [inline examples/mechanics/fragmenting_cylinder.cpp:61]: We need to ensure this example works with LPS.

File: src/CabanaPD_Boundary.hpp [hunk 12]
```diff
@@ -40,14 +52,53 @@ struct RegionBoundary
...
+    KOKKOS_INLINE_FUNCTION bool compare( const PositionType& x,
+                                         const int pid ) const
+    {
+        return ( x( pid, 0 ) >= low_x && x( pid, 0 ) <= high_x &&
+                 x( pid, 1 ) >= low_y && x( pid, 1 ) <= high_y &&
+                 x( pid, 2 ) >= low_z && x( pid, 2 ) <= high_z );
+    }
+};
+
+// Define a cylinder as the boundary.
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:66]: Note in this example we are applying the boundary condition everywhere in the domain.

File: src/CabanaPD_Types.hpp [hunk 13]
```diff
@@ -40,5 +40,9 @@ struct LinearLPS
 {
 };
 
+struct DynamicCriticalStretch
```
Comment (pabloseleson) [inline src/CabanaPD_Types.hpp:43]: We may move the DynamicCriticalStretch to a different PR.

File: src/force/CabanaPD_ForceModels_PMB.hpp (hunk 9 above)
Comment (streeve) [reply src/force/CabanaPD_ForceModels_PMB.hpp:299]: No

File: examples/mechanics/fragmenting_cylinder.cpp (hunk 11 above)
Comment (streeve) [reply examples/mechanics/fragmenting_cylinder.cpp:61]: Need to merge #116

File: src/CabanaPD_Types.hpp (hunk 13 above)
Comment (streeve) [reply src/CabanaPD_Types.hpp:43]: Opened in #117

File: examples/mechanics/fragmenting_cylinder.cpp (hunk 10 above)
Comment (streeve) [reply examples/mechanics/fragmenting_cylinder.cpp:86]: Need to merge #114

File: src/CabanaPD_Boundary.hpp (hunk 5 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp]: Removed from this PR

File: src/CabanaPD_Boundary.hpp [hunk 14]
```diff
@@ -40,14 +52,53 @@ struct RegionBoundary
...
+    KOKKOS_INLINE_FUNCTION bool compare( const PositionType& x,
+                                         const int pid ) const
+    {
+        return ( x( pid, 0 ) >= low_x && x( pid, 0 ) <= high_x &&
+                 x( pid, 1 ) >= low_y && x( pid, 1 ) <= high_y &&
+                 x( pid, 2 ) >= low_z && x( pid, 2 ) <= high_z );
+    }
+};
+
+// Define a cylinder or other radial subset of the system as the boundary.
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/CabanaPD_Boundary.hpp]: Removed from this PR

File: examples/mechanics/fragmenting_cylinder.cpp [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        // Density
+        rho( pid ) = rho0;
+
+        // Perturb particle positions
+        double random_number_x = ( (double)rand() / ( RAND_MAX ) );
+        double random_number_y = ( (double)rand() / ( RAND_MAX ) );
+        double random_number_z = ( (double)rand() / ( RAND_MAX ) );
+
+        double factor = 0.001; // inputs["grid_perturbation_factor"];
```
Comment (pabloseleson) [inline examples/mechanics/fragmenting_cylinder.cpp:122]: When trying to use 

double factor = inputs["grid_perturbation_factor"];

I got the error below. Can you fix this?

CabanaPD/examples/mechanics/fragmenting_cylinder.cpp:123:31: error: no viable overloaded operator[] for type 'const CabanaPD::Inputs'
        double factor = inputs["grid_perturbation_factor"];
                        ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
CabanaPD/src/CabanaPD_Input.hpp:185:10: note: candidate function not viable: 'this' argument has type 'const CabanaPD::Inputs', but method is not marked const
    auto operator[]( std::string label ) { return inputs[label]["value"]; }
         ^
1 error generated.
make[2]: *** [examples/mechanics/CMakeFiles/FragmentingCylinder.dir/fragmenting_cylinder.cpp.o] Error 1
make[1]: *** [examples/mechanics/CMakeFiles/FragmentingCylinder.dir/all] Error 2
make: *** [all] Error 2

Comment (streeve) [reply examples/mechanics/fragmenting_cylinder.cpp:122]: You can't call that inside the kernel - move it before and you can still use it inside

Comment (pabloseleson) [reply examples/mechanics/fragmenting_cylinder.cpp]: Done
